pkg/controller/certificatesigningrequests/acme/acme.go

```diff
@@ -0,0 +1,308 @@
+	cmacmeclientset "github.com/jetstack/cert-manager/pkg/client/clientset/versioned/typed/acme/v1"
+	cmacmelisters "github.com/jetstack/cert-manager/pkg/client/listers/acme/v1"
+	controllerpkg "github.com/jetstack/cert-manager/pkg/controller"
+	"github.com/jetstack/cert-manager/pkg/controller/certificatesigningrequests"
+	ctrlutil "github.com/jetstack/cert-manager/pkg/controller/certificatesigningrequests/util"
+	logf "github.com/jetstack/cert-manager/pkg/logs"
+	"github.com/jetstack/cert-manager/pkg/util"
+	"github.com/jetstack/cert-manager/pkg/util/pki"
+)
+
+const (
+	CSRControllerName = "certificatesigningrequests-issuer-acme"
+)
+
+// ACME is a Kubernetes CertificateSigningRequest controller, responsible for
+// signing CertificateSigningRequests that reference a cert-manager ACME Issuer
+// or ClusterIssuer
+type ACME struct {
+	issuerOptions controllerpkg.IssuerOptions
+
+	orderLister cmacmelisters.OrderLister
+	acmeClientV cmacmeclientset.AcmeV1Interface
+	certClient  certificatesclient.CertificateSigningRequestInterface
+
+	recorder record.EventRecorder
+}
+
+func init() {
+	// create certificate request controller for acme issuer
+	controllerpkg.Register(CSRControllerName, func(ctx *controllerpkg.Context) (controllerpkg.Interface, error) {
+		return controllerpkg.NewBuilder(ctx, CSRControllerName).
+			For(certificatesigningrequests.New(apiutil.IssuerACME, NewACME(ctx), ctx.SharedInformerFactory.Acme().V1().Orders().Informer())).
+			Complete()
+	})
+}
+
+func NewACME(ctx *controllerpkg.Context) *ACME {
+	return &ACME{
+		issuerOptions: ctx.IssuerOptions,
+		orderLister:   ctx.SharedInformerFactory.Acme().V1().Orders().Lister(),
+		acmeClientV:   ctx.CMClient.AcmeV1(),
+		certClient:    ctx.Client.CertificatesV1().CertificateSigningRequests(),
+		recorder:      ctx.Recorder,
+	}
+}
+
+// Sign attempts to sign the given CertificateSigningRequest based on the
+// provided ACME Issuer or ClusterIssuer.
+//
+// If no order exists for a CertificateSigningRequest, an order is constructed
+// and sent back to the Kubernetes API server for processing.  The order
+// controller then processes the order. The CertificateSigningRequest is then
+// updated with the result.
+func (a *ACME) Sign(ctx context.Context, csr *certificatesv1.CertificateSigningRequest, issuerObj cmapi.GenericIssuer) error {
+	log := logf.FromContext(ctx, "sign")
+
+	// If we can't decode the CSR PEM we have to hard fail
+	req, err := pki.DecodeX509CertificateRequestBytes(csr.Spec.Request)
+	if err != nil {
+		message := fmt.Sprintf("Failed to decode CSR in spec.request: %s", err)
+		log.Error(err, message)
+		a.recorder.Event(csr, corev1.EventTypeWarning, "RequestParsingError", message)
+		ctrlutil.CertificateSigningRequestSetFailed(csr, "RequestParsingError", message)
+		_, err = a.certClient.UpdateStatus(ctx, csr, metav1.UpdateOptions{})
+		return err
+	}
+
+	// If the CommonName is also not present in the DNS names or IP Addresses of the Request then hard fail.
```

Comment:
nit: inconsistent comment wrapping width throughout the file. Sometimes 80 chars, sometimes 100 chars; keeping it under 100 chars is nice to help readability on GitHub PRs, and 80 chars are nice for making sure GitHub inline suggestions are nicely rendered. Just keep the hard wraps consistent throughout the file 😅

tip: `vim` can hard rewrap comments with `gqq` on a visual selection.